[task/ASV-1800] Refactor AppView load app

franciscoaleixo opened the pull request:
**What does this PR do?**

   This PR refactors the initial load of the app view on the presenter side. As a side effect, many other things that are listed below were refactored. This PR came because of visual issues with the new DownloadProgressView that occured as a result of the mess that was loadApp().

List of refactored things:
1. Major changes to AppViewModel.
   1. Renamed old AppViewModel to AppModel. This contains the mapping from DetailedApp.
   2. New AppViewModel class. This class contains other Models that are essential to the AppView and used in different components. This includes AppModel, DownloadModel, AppCoinsViewModel and MigrationModel. This also makes it easy to add new models if need be.
   3. New AppViewModelManager class. The new AppViewModel is managed by this new Manager, which is responsible for caching relevant data and efficiently providing AppViewModel.
2. Major flow changes to the loadApp() chain.
   1. Un-nested and separated each action to its own type. What this means is that every view operation is together, every analytic operation is together, etc. This makes it very easy to follow what's happening and where to add relevant code.
   2. Extracted methods to self-document code.
   3. Refactored AppView open types to avoid duplicated code. Fixed bug with OPEN_AND_INSTALL where the action was hardcoded.

Among other minor things.

**Database changed?**

   No

**Where should the reviewer start?**

- [ ] AppViewPresenter.java
- [ ] AppViewModel.java

**How should this be manually tested?**

  There's a lot of changes to the appview presenter/manager code, so it needs to be tested thoroughly. Just try to break the appview in general, but the main focus is the initialization. Test downloads, check appc apps, etc. Don't forget to test Appc Promotions because there were some changes there as well (check https://github.com/Aptoide/aptoide-client-v8/pull/1016  to see how to test).

**What are the relevant tickets?**

  [ASV-1800](https://aptoide.atlassian.net/browse/ASV-1800)


**Code Review Checklist**

- [ ] Documentation on public interfaces
- [ ] Database changed? If yes - Migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] New Kotlin code has unit tests
- [ ] New flows in presenters unit tests
- [ ] Mappers/Validators with any kind of logic unit tests
- [ ] Functional tests pass

franciscoaleixo commented:
Tests finalized, can be fully reviewed 👍

jdandrade commented on app/src/main/java/cm/aptoide/pt/app/view/AppViewPresenter.java:
```diff
@@ -138,6 +138,96 @@ public AppViewPresenter(AppViewView view, AccountNavigator accountNavigator,
+    DownloadModel.Action action = appViewModel.getDownloadModel()
+        .getAction();
+    return accountManager.accountStatus()
+        .first()
```
this accountStatus().first() not necessary. Probably it came from older code.

jdandrade commented on app/src/main/java/cm/aptoide/pt/app/view/AppViewPresenter.java:
```diff
@@ -138,6 +138,96 @@ public AppViewPresenter(AppViewView view, AccountNavigator accountNavigator,
+        .first()
+        .observeOn(viewScheduler)
+        .flatMap(__ -> handleOpenAppViewDialogInput(appViewModel.getAppModel()))
+        .filter(shouldDownload -> shouldDownload)
```
In a normal click to open app view with OPEN_ONLY OpenType, shouldDownload will be false and nothing else in the chain will happen, which it shouldn't for the downloadApp and so on but when you go back to the loadAppView() the loadOtherAppViewComponents(appModel)  method isn't called. This should be the reason why the similar apps, reviews aren't being shown. The banner, legacy ads and donations should also not show up because of this, but didn't test.

franciscoaleixo replied:
This should now be fixed 👍 I've added tests to verify that these methods are being called so that this won't happen again.

franciscoaleixo commented:
I made a lot of new changes, sorry. I think it was worth it though, let me know what you think!

- Removed 'view.isAppReadyReadyToDownload'. This just really was a check for cached AppViewModel and it basically smelled like a hack. For places where this was used with downloadInRange, it was replaced with view.installAppClick() + downloadInRange (since those generally make sense when an install is clicked anyway). For other places, it was simply added to the loadAppView() method, since they really were just initializing methods.
- Filter for AppModel error is now earlier in the loadAppView() stream to avoid doing unnecessary work if the AppModel wasn't loaded.
- Initializing streams are now merged with 'mergeDelayError' instead of running sequentially. This has two benefits: it runs the initializing streams at the same time, improving loading time of those streams slightly, and errors in one stream don't affect the other streams (hence the 'delayError'). Every stream should still handle errors appropriately if they need to, though.
 - Organic ads and banner ads that were initialized ONLY on firstLoad are now on loadAppView (because it should work when a user clicks retry as well!). With this they were slightly refactored to match current implementation.

@jdandrade

jdandrade commented:
Another issue that is happening is in APPC app views, when you go to another view (read more for example) and go back, the app view always scrolls up to the install button (you don't see scrolling animation, it just restores state on top of the install button.

jdandrade commented on aptoide-analytics-core/src/main/java/cm/aptoide/analytics/AnalyticsManager.java:
```diff
@@ -36,12 +36,8 @@ private AnalyticsManager(KnockEventLogger knockLogger,
-        + data
-        + "],  action = ["
+    analyticsLogger.logDebug(TAG, "logEvent() called with: " + "eventName = ["
+        + eventName + "], data = [" + data + "],  action = ["
```
codestyle

jdandrade commented:
> Clicking install > pause > go to another app view (Similar apps for example) > back > resume doesn't work.
> And seems like something funky might be going on with the analytics. Are we sending stuff on each tick of the download progress?
> Everything else was solved and seems more robust now. Will keep checking this (more functional testing right now, architecture seems right)

Pause > Resuming Promotion Progress is also not working. Don't even need to leave the view. Not sure if this is somewhat related.

franciscoaleixo commented on app/src/main/java/cm/aptoide/pt/app/AppViewManager.java:
```diff
@@ -441,17 +343,12 @@ public void handleAdsLogic(SearchAdResult searchAdResult) {
+    return getAppModel().doOnSuccess(appModel -> {
+      appModel.getPay()
           .setPaid();
-      cachedApp.setPath(path);
```
Yeah I removed it at the time, because it seemed like it wasn't doing anything but seems like I was wrong :<... Since we will be removing support it probably won't matter much but I'll add it back.

franciscoaleixo commented:
@jdandrade About the resume... So it's pretty much broken but I think it's not related with this refactor but rather another bug that was being worked on by @phillipsgoncalves. I added an error catch to the resume chain and it says 

```
W/System.err: java.lang.NullPointerException: Attempt to invoke virtual method 'int cm.aptoide.pt.app.DownloadModel$Action.ordinal()' on a null object reference
W/System.err:     at cm.aptoide.pt.app.AppViewAnalytics.mapDownloadAction(AppViewAnalytics.java:321)
W/System.err:     at cm.aptoide.pt.app.AppViewAnalytics.setupDownloadEvents(AppViewAnalytics.java:312)
```

Which, if I'm not mistaken, has to do with the download events. Can you confirm @phillipsgoncalves to see if it's me or not?

franciscoaleixo commented:
@jdandrade After talking with @phillipsgoncalves , the issue with resume has to do with https://github.com/Aptoide/aptoide-client-v8/pull/1072. With his help, it's now fixed in this branch with the new AppViewModel. This should be also fixed for the promotion resume.

About the restore scroll. Unfortunately I can't replicate this. But I didn't test with an Appc Promotions. Did you have it when it occured? I think it's because the appc promotion is loaded after, so it might interfere with the scroll restore. Can you check that this is the case? I'm not sure if we can do anything about it if that's the case. We could try to restore only after the promotions load, but it might be janky if it takes a while to load...

I've also added the path change to paid apps back.

franciscoaleixo commented:
I made an additional micro-optimization that is pretty cool. So normally, doing a getAppViewModel() with cached AppModel takes from around 7 to +30 ms. This time is about 99% spent on retrieving the DownloadModel, which is not cached. This time is "small", but it builds up with the loading time of other components that rely on this as an entry point.

However, we know that we have a stream always listening on this (observeDownloadApp() on AppViewPresenter, for example). So it would be really nice if we could safely share the latest event of this stream with any other stream that needs AppViewModel. Well, we could if we were to combine replay(1), share() and autoconnect() operators. Unfortunately they don't play well together, however Jake Wharton has a small operator just for this: RxJava Replaying Share. Also, this will never have stale cached data, because if every subscriber unsubscribes, it will discard the cached data.

This drops the getAppViewModel() time to <0ms - 1 ms. It speeds up every operation that needs this and improves loading times for AppView on backstack.

Some random benchmarks (far from scientific lol). I promise this is the last change I will do to this PR too!
```
# WITHOUT ReplayingShare
# First getAppViewModel time is without AppModel cached
2019-07-20 00:34:01.556 27685-27896/cm.aptoide.pt.dev I/getAppViewModel_time: 762
2019-07-20 00:34:02.207 27685-27896/cm.aptoide.pt.dev I/getAppViewModel_time: 14
2019-07-20 00:34:02.215 27685-27906/cm.aptoide.pt.dev I/getAppViewModel_time: 11
2019-07-20 00:34:03.889 27685-27906/cm.aptoide.pt.dev I/getAppViewModel_time: 10
2019-07-20 00:34:10.037 27685-27905/cm.aptoide.pt.dev I/getAppViewModel_time: 11
2019-07-20 00:34:10.699 27685-27864/cm.aptoide.pt.dev I/getAppViewModel_time: 19
2019-07-20 00:34:10.704 27685-27820/cm.aptoide.pt.dev I/getAppViewModel_time: 21
2019-07-20 00:34:10.711 27685-27904/cm.aptoide.pt.dev I/getAppViewModel_time: 21
2019-07-20 00:34:13.678 27685-27903/cm.aptoide.pt.dev I/getAppViewModel_time: 10
2019-07-20 00:34:14.304 27685-27895/cm.aptoide.pt.dev I/getAppViewModel_time: 13
2019-07-20 00:34:14.312 27685-27899/cm.aptoide.pt.dev I/getAppViewModel_time: 12
2019-07-20 00:34:14.318 27685-27882/cm.aptoide.pt.dev I/getAppViewModel_time: 15
2019-07-20 00:34:17.088 27685-27906/cm.aptoide.pt.dev I/getAppViewModel_time: 9
2019-07-20 00:34:17.716 27685-27837/cm.aptoide.pt.dev I/getAppViewModel_time: 9
2019-07-20 00:34:17.722 27685-27901/cm.aptoide.pt.dev I/getAppViewModel_time: 16
2019-07-20 00:34:17.727 27685-27895/cm.aptoide.pt.dev I/getAppViewModel_time: 11
2019-07-20 00:34:20.473 27685-27905/cm.aptoide.pt.dev I/getAppViewModel_time: 7
2019-07-20 00:34:21.131 27685-27864/cm.aptoide.pt.dev I/getAppViewModel_time: 12
2019-07-20 00:34:21.137 27685-27836/cm.aptoide.pt.dev I/getAppViewModel_time: 14
2019-07-20 00:34:21.142 27685-27901/cm.aptoide.pt.dev I/getAppViewModel_time: 15
2019-07-20 00:34:23.455 27685-27898/cm.aptoide.pt.dev I/getAppViewModel_time: 10
2019-07-20 00:34:24.077 27685-27903/cm.aptoide.pt.dev I/getAppViewModel_time: 16
2019-07-20 00:34:24.090 27685-27901/cm.aptoide.pt.dev I/getAppViewModel_time: 24
2019-07-20 00:34:24.098 27685-27905/cm.aptoide.pt.dev I/getAppViewModel_time: 31

# WITH ReplayingShare
2019-07-20 00:36:08.214 28079-28261/cm.aptoide.pt.dev I/getAppViewModel_time: 825
2019-07-20 00:36:08.845 28079-28268/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:08.847 28079-28303/cm.aptoide.pt.dev I/getAppViewModel_time: 0
2019-07-20 00:36:10.480 28079-28238/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:29.061 28079-28079/cm.aptoide.pt.dev I/getAppViewModel_time: 0
2019-07-20 00:36:29.689 28079-28303/cm.aptoide.pt.dev I/getAppViewModel_time: 0
2019-07-20 00:36:29.708 28079-28268/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:29.712 28079-28303/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:32.590 28079-28079/cm.aptoide.pt.dev I/getAppViewModel_time: 0
2019-07-20 00:36:33.215 28079-28238/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:33.216 28079-28285/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:33.228 28079-28268/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:34.848 28079-28079/cm.aptoide.pt.dev I/getAppViewModel_time: 0
2019-07-20 00:36:35.479 28079-28267/cm.aptoide.pt.dev I/getAppViewModel_time: 0
2019-07-20 00:36:35.480 28079-28297/cm.aptoide.pt.dev I/getAppViewModel_time: 0
2019-07-20 00:36:35.481 28079-28268/cm.aptoide.pt.dev I/getAppViewModel_time: 3
2019-07-20 00:36:37.754 28079-28079/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:38.369 28079-28303/cm.aptoide.pt.dev I/getAppViewModel_time: 0
2019-07-20 00:36:38.374 28079-28268/cm.aptoide.pt.dev I/getAppViewModel_time: 1
2019-07-20 00:36:38.374 28079-28297/cm.aptoide.pt.dev I/getAppViewModel_time: 1
```

jdandrade commented:
Migration Tests:
   Migration Promotion > Click Wallet Install > pause at the beggining 5-50% > click update to APPC > pause > resume wallet install > no progress showing up but app is being downloaded and install is prompt.
   "Update to appcoins" > click > pause > cancel > button goes "INSTALL" instead of Update to AppCoins (this should be failing in production already...)
   Cancel Wallet install ( in promotion) doesn't seem to do anything at the moment when wallet is 100% downloaded.

jdandrade commented:
Sorry, I wanted to accept this earlier and now it has conflicts... If you can solve @franciscoaleixo... I will accept immediately when solved.
